server/chat-plugins/abuse-monitor.ts

```diff
@@ -0,0 +1,715 @@
+/**
+ * Neural net chat'filters'.
+ * These are in a separate file so that they don't crash the other filters.
+ * (issues with globals, etc)
+ * We use Google's Perspective API to classify messages.
+ * @see https://perspectiveapi.com/
+ * by Mia.
+ * @author mia-pi-git
+ */
+
+import {FS, Utils, Net, ProcessManager, Repl} from '../../lib';
+import {Config} from '../config-loader';
+import {toID} from '../../sim/dex-data';
+
+const WHITELIST = ["mia"];
+// 20m. this is mostly here so we can use Monitor.slow()
+const PM_TIMEOUT = 20 * 60 * 1000;
+const ATTRIBUTES = {
+	"SEVERE_TOXICITY": {},
+	"TOXICITY": {},
+	"IDENTITY_ATTACK": {},
+	"INSULT": {},
+	"PROFANITY": {},
+	"THREAT": {},
+};
+
+export const cache: {
+	[roomid: string]: {[userid: string]: number} & {
+		staffNotified?: boolean,
+		claimed?: ID,
+	},
```

Comment:
Why is this `&` rather than a `users` sub-object?

Reply:
I thought it was more readable, personally.

Reply:
Nope, that's actually necessary for TS too.

Reply:
Discussed in dms, restored cache[roomid].users.